### PR TITLE
feat(*): add environment variable, tty and random functions

### PR DESCRIPTION
**Pull request by Tieske**
Adds;
- environment variable (get, set, list)
- tty; isatty
- random: high quality random data

Additionally rewrites documentation and support files (chaneglog, readme, license) for inclusion in generated docs.

### CODE REVIEW

**Comment by Tieske** [comment]
@o-lim this PR is the final result. Please have a look. Generated docs are in `/docs`.

File: src/random.c [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+#ifdef _WIN32
+#include "windows.h"
+#include "wincrypt.h"
+#else
+#include <unistd.h>
+#endif
+
+
+// Maximum buffer size for random bytes
+#define MAX_RANDOM_BUFFER_SIZE 1024
```
Comment (o-lim) [inline src/random.c:16]: Should there be a limit? If there is a limit, should it be configurable from lua?

File: src/random.c [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+Generate random bytes.
+This uses `getrandom()` on Linux, `CryptGenRandom()` on Windows, and `/dev/urandom` on macOS.
+@function random
+@tparam[opt=1] int length number of bytes to get, must be less than or equal to `MAX_RANDOM_BUFFER_SIZE` (1024)
+@treturn[1] string string of random bytes
+@treturn[2] nil
+@treturn[2] string error message
+*/
+static int lua_get_random_bytes(lua_State* L) {
+    int num_bytes = luaL_optinteger(L, 1, 1); // Number of bytes, default to 1 if not provided
```
Comment (o-lim) [inline src/random.c:29]: Should the `getrandom` function also take another optional argument specify a truly random source (i.e. `/dev/random`) vs the default pseudorandom source (i.e. `/dev/urandom`)?

File: src/environment.c [hunk 3]
```diff
@@ -12,8 +12,8 @@
 /***
 Gets the value of an environment variable.
 
-__NOTE__: Windows has multiple copies of environment variables. This function
-will work with Lua's `setenv` on Windows. If you want to use `setenv` then
+__NOTE__: Windows has multiple copies of environment variables. The `setenv` function
+will not work with Lua's `os.getenv` on Windows. If you want to use `setenv` then
```
Comment (o-lim) [inline src/environment.c:16]: Should this change be in a prior commit where `environment.c` was added?

This comment should also describe why it won't work with Lua's `setenv` on Windows, and perhaps re-enforce that it will work for non-Windows. Is the reason it won't work for Windows due to the multiple copies? If so, then maybe rephrase to
```
__NOTE__: Windows has multiple copies of environment variables. For this reason, the `setenv` function
will not work with Lua's `os.getenv` on Windows. ...
```

Comment (Tieske) [reply src/environment.c:16]: > Should this change be in a prior commit where environment.c was added?

It probably should have been, I messed it up a little. Let me know if you want me to move it around.

fixed the comment

File: src/random.c (hunk 2 above)
Comment (Tieske) [reply src/random.c:29]: `/dev/random` can block. Hence didn't use that. It is available only on unixes anyway, and there one could use pure-lua to read from the file already.

So the idea is to have `result, err = getrandom(size, no_u)` where `no_u` is a boolean that forces reading from `/dev/random` if set to `true` ?

File: src/random.c (hunk 1 above)
Comment (Tieske) [reply src/random.c:16]: Any usecases for needing more than 1024 bytes at once? can easily be solved on Lua side (one even has to, because the generators are not guaranteeing the full number requested to be returned).

Comment (Tieske) [reply src/random.c:16]: this actually had a bug, now fixed.

File: src/random.c (hunk 2 above)
Comment (o-lim) [reply src/random.c:30]: > So the idea is to have `result, err = getrandom(size, no_u)` where `no_u` is a boolean that forces reading from `/dev/random` if set to true ?

Yeah, pretty much. However, if a truly random source is not available on Windows, then maybe it's not worth it. We would want to default to pseudorandom (i.e. `/dev/urandom`) since `/dev/random` can block. It's fine if it's just pseudorandom, which doesn't seem like a big deal if truly random is not available on all platforms.

File: src/environment.c (hunk 3 above)
Comment (o-lim) [reply src/environment.c]: Could you fix it in the original commit then, just so there's a cleaner history?

File: src/random.c (hunk 1 above)
Comment (o-lim) [reply src/random.c:16]: I think this should work similar to `file:read()`, which does not have a limit either
```
file:read (···)
    The available formats are 
        ...
        * number: reads a string with up to this number of bytes, returning nil on end of file. If number is zero, it reads nothing and returns an empty string, or nil on end of file. 
```

If the number of request bytes is greater than `RANDOM_BUFFER_SIZE (1024)`, we can allocate it from the heap rather than the stack, and fill in the buffer. Or just allocate from the heap regardless. Either way works. If allocation fails, just fail with a `ENOMEM` error.

Comment (Tieske) [reply src/random.c:16]: How about we just cap it at 1024, instead of returning an error? As the user will always have to check the return length anyway and retry if not enough. Keeps the implementation simple, and has no extra complexity in Lua land.

File: src/environment.c (hunk 3 above)
Comment (Tieske) [reply src/environment.c]: will cleanup once the other comments are settled.

File: src/random.c (hunk 1 above)
Comment (Tieske) [reply src/random.c]: done

File: src/random.c [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        lua_pushfstring(L, "failed to get random data: %lu", error);
+        CryptReleaseContext(hCryptProv, 0);
+        return 2;
+    }
+
+    CryptReleaseContext(hCryptProv, 0);
+#else
+#ifndef __APPLE__
+    // Neither Apple nor Windows
+    n = getrandom(buffer, num_bytes, 0);
```
[the remark this answers is not in the document]
Comment (Tieske) [reply src/random.c]: done

[the remark this answers is not in the document]
Comment (Tieske) [reply src/random.c]: done

File: src/random.c [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+@treturn[1] string string of random bytes
+@treturn[2] nil
+@treturn[2] string error message
+*/
+static int lua_get_random_bytes(lua_State* L) {
+    int num_bytes = luaL_optinteger(L, 1, 1); // Number of bytes, default to 1 if not provided
+
+    if (num_bytes <= 0) {
+        lua_pushnil(L);
+        lua_pushstring(L, "invalid number of bytes, must be greater than 0");
```
Comment (o-lim) [inline src/random.c:32]: Since `getrandom(buffer, 0, 0)` does not return an error, I don't think requesting 0 bytes should be an error. After all, it's still doing what you told it to do, which is generate 0 random bytes. It should return the empty string if `num_bytes == 0`.

Comment (Tieske) [reply src/random.c]: done!

**Comment by Tieske** [comment]
thx for the thorough review 👍 . My c skillz are rustier than I thought 🙈 .

File: src/random.c [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        lua_pushfstring(L, "failed to get random data: %lu", error);
+        CryptReleaseContext(hCryptProv, 0);
+        return 2;
+    }
+
+    CryptReleaseContext(hCryptProv, 0);
+#else
+
+    // for macOS/unixes use /dev/urandom for non-blocking
+    int fd = open("/dev/urandom", O_CLOEXEC);
```
Comment (o-lim) [inline src/random.c:71]: This should be `O_RDONLY | O_CLOEXEC`

File: src/term.c [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+/***
+Checks if a file-handle is a TTY.
+
+@function isatty
+@tparam file file the file-handle to check
+@treturn boolean true if the file is a tty
+*/
+static int lua_isatty(lua_State* L) {
+    FILE **fh = (FILE **) luaL_checkudata(L, 1, LUA_FILEHANDLE);
+    lua_pushboolean(L, isatty(fileno(*fh)));
```
Comment (o-lim) [inline src/term.c:21]: I believe `isatty` is already part of [lua-term](https://github.com/hoelzro/lua-term), which also provides other terminal functions. Do you think it needs to be included in this module as well?

Comment (Tieske) [reply src/term.c:21]: yes, the one in lua-term is actually deprecated, pointing to lua-posix as a successor (which is a massive kitchen-sink library).

File: src/compat.h [hunk 8]
```diff
@@ -8,4 +8,31 @@
...
+#elif SIZE_MAX == ULLONG_MAX
+typedef long long ssize_t;  /* windows 64 bits */
+#define SSIZE_MIN  LLONG_MIN
+#define SSIZE_MAX  LLONG_MAX
+#elif SIZE_MAX == USHRT_MAX
+typedef short ssize_t;      /* is this even possible? */
+#define SSIZE_MIN  SHRT_MIN
+#define SSIZE_MAX  SHRT_MAX
+#elif SIZE_MAX == UINTMAX_MAX
+typedef uintmax_t ssize_t;  /* last resort, chux suggestion */
```
Comment (o-lim) [inline src/compat.h:30]: `ssize_t` should be signed, so should this be `intmax_t`?

Comment (Tieske) [reply src/compat.h:30]: this is only used on Windows for the absence of ssize_t, which should end up in the "common 32bit" or "windows 64bit" branches.
Source: https://stackoverflow.com/a/34580624/746559

Comment (o-lim) [reply src/compat.h:30]: I understand that it's Windows only, but it still looks wrong. `ssize_t` should always be signed, not unsigned. Seems like a typo in the original post.

Comment (Tieske) [reply src/compat.h:30]: yep thats definitely true

Comment (Tieske) [reply src/compat.h]: fixed!

**Review by o-lim — APPROVED** [review]

**Comment by Tieske** [comment]
:tada: Merged, all CI green, and docs are up! Thx @o-lim !
